tmt/lint.py

```diff
@@ -0,0 +1,308 @@
+        """
+
+        for name in dir(cls):
+            if not name.startswith('lint_'):
+                continue
+
+            cls.get_linter_registry().append(tmt.lint.Linter(getattr(cls, name)))
+
+    @classmethod
+    def resolve_enabled_linters(
+            cls,
+            enable_checks: Optional[List[str]] = None,
+            disable_checks: Optional[List[str]] = None
+            ) -> List[Linter]:
+        linters: List[Linter] = []
+
+        if not enable_checks:
+            linters = cls.get_linter_registry()
+
+        else:
+            linters = []
+
+            for needle in enable_checks:
+                linters += [
+                    linter
+                    for linter in cls.get_linter_registry()
+                    if needle in linter.id
+                    ]
+
+        if disable_checks:
+            linters = [
+                linter for linter in linters
+                if linter.id not in disable_checks
+                ]
+
+        return linters
+
+    def lint(
+        self,
+        enable_checks: Optional[List[str]] = None,
+        disable_checks: Optional[List[str]] = None,
+        enforce_checks: Optional[List[str]] = None
+            ) -> Tuple[bool, List[LinterRuling]]:
+        """
+        Check the instance against a battery of linters and report results.
+
+        :returns: whether the instance is valid.
+        """
+
+        enforce_checks = enforce_checks or []
+
+        linters = self.resolve_enabled_linters(
+            enable_checks=enable_checks,
+            disable_checks=disable_checks)
+
+        valid = True
+        rulings: List[LinterRuling] = []
+
+        for linter in sorted(linters, key=lambda x: x.id):
+            for outcome, message in linter.callback(self):
+                if outcome == LinterOutcome.FAIL:
+                    rulings.append((linter, outcome, outcome, message))
+
+                    valid = False
+
+                elif outcome != LinterOutcome.PASS:
+                    if linter.id in enforce_checks:
+                        rulings.append((linter, outcome, LinterOutcome.FAIL, message))
+
+                        valid = False
+
+                    else:
+                        rulings.append((linter, outcome, outcome, message))
+
+                else:
+                    rulings.append((linter, outcome, outcome, message))
+
+        return valid, rulings
+
+    @classmethod
+    def format_linters(cls) -> str:
+        hints: List[str] = []
+
+        for linter in sorted(cls.get_linter_registry(), key=lambda x: x.id):
+            # Thsi has been already verified when registering the linter
+            assert linter.callback.__doc__ is not None
+
+            hints.append(textwrap.dedent(linter.callback.__doc__).strip())
+
+        return '\n'.join(hints)
+
+
+def filter_allowed_checks(
+        rulings: Iterable[LinterRuling],
+        outcomes: Optional[List[LinterOutcome]] = None) -> Generator[LinterRuling, None, None]:
```

Comment:
you already started with return type on new line, it might look better to keep that style here as well
```suggestion
        outcomes: Optional[List[LinterOutcome]] = None
        ) -> Generator[LinterRuling, None, None]:
```